istioctl/cmd/kubeuninject.go

```diff
@@ -0,0 +1,373 @@
+
+package cmd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+
+	"github.com/ghodss/yaml"
+	openshiftv1 "github.com/openshift/api/apps/v1"
+	"github.com/spf13/cobra"
+	"go.uber.org/multierr"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/api/batch/v2alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	proxyContainerName              = "istio-proxy"
+	initContainerName               = "istio-init"
+	envoyVolumeName                 = "istio-envoy"
+	certVolumeName                  = "istio-certs"
+	annotationPolicy                = "sidecar.istio.io/inject"
+	annotationStatus                = "sidecar.istio.io/status"
+	annotationRewriteAppHTTPProbers = "sidecar.istio.io/rewriteAppHTTPProbers"
+)
+
+var (
+	uninjectKinds = []struct {
+		groupVersion schema.GroupVersion
+		obj          runtime.Object
+		resource     string
+		apiPath      string
+	}{
+		{corev1.SchemeGroupVersion, &corev1.ReplicationController{}, "replicationcontrollers", "/api"},
+		{corev1.SchemeGroupVersion, &corev1.Pod{}, "pods", "/api"},
+
+		{appsv1.SchemeGroupVersion, &appsv1.Deployment{}, "deployments", "/apis"},
+		{appsv1.SchemeGroupVersion, &appsv1.DaemonSet{}, "daemonsets", "/apis"},
+		{appsv1.SchemeGroupVersion, &appsv1.ReplicaSet{}, "replicasets", "/apis"},
+
+		{batchv1.SchemeGroupVersion, &batchv1.Job{}, "jobs", "/apis"},
+		{v2alpha1.SchemeGroupVersion, &v2alpha1.CronJob{}, "cronjobs", "/apis"},
+
+		{appsv1.SchemeGroupVersion, &appsv1.StatefulSet{}, "statefulsets", "/apis"},
+
+		{corev1.SchemeGroupVersion, &corev1.List{}, "lists", "/apis"},
+
+		{openshiftv1.GroupVersion, &openshiftv1.DeploymentConfig{}, "deploymentconfigs", "/apis"},
+	}
+	uninjectScheme = runtime.NewScheme()
+)
+
+func init() {
+	for _, kind := range uninjectKinds {
+		uninjectScheme.AddKnownTypes(kind.groupVersion, kind.obj)
+		uninjectScheme.AddUnversionedTypes(kind.groupVersion, kind.obj)
+	}
+}
+
+func validateUninjectFlags() error {
+	var err error
+
+	if uninjectInFilename == "" {
+		err = multierr.Append(err, errors.New("filename not specified (see --filename or -f)"))
+	}
+	return err
+}
+
+func fromRawToObject(raw []byte) (runtime.Object, error) {
+	var typeMeta metav1.TypeMeta
+	if err := yaml.Unmarshal(raw, &typeMeta); err != nil {
+		return nil, err
+	}
+
+	gvk := schema.FromAPIVersionAndKind(typeMeta.APIVersion, typeMeta.Kind)
+	obj, err := uninjectScheme.New(gvk)
+	if err != nil {
+		return nil, err
+	}
+	if err = yaml.Unmarshal(raw, obj); err != nil {
+		return nil, err
+	}
+
+	return obj, nil
+}
+
+// extractResourceFile uninjects the istio proxy into the specified
+// kubernetes YAML file.
+func extractResourceFile(in io.Reader, out io.Writer) error {
+	reader := yamlDecoder.NewYAMLReader(bufio.NewReaderSize(in, 4096))
```

Comment:
Can you please add some check here that if the input file is the in the original version( uninjected) then simply skip all the following process?